docs/architecture/adr-002-ibc-relayer.md

```diff
@@ -0,0 +1,644 @@
+
+#### Batching
+The relayer may batch transactions if supported by destination chain and allowed by configuration. In this case the relayer  can bundle multiple datagrams into a single transaction and amortise any overhead costs (e.g. signature checks for fee payment).
+Initial version of the relayer assumes batching is supported by all chains. An option may be later included in the configuration file.
+
+## Relayer Requirements
+
+A correct relayer MUST:
+
+- **[R-config-start]** Read, parse, validate a configuration file upon start and configure itself for the specifed chains and paths
+- **[R-transport]** Have access to the networking protocols (e.g. TCP/IP, UDP/IP, or QUIC/IP) and physical transport, required to read the state of one blockchain/ machine and submit data to another
+- **[R-provider]** Maintain transport connections to at least one full node per chain
+- **[R-query]** Query IBC data on source and destination chains
+- **[R-light-client]** Run light clients for source chains and 
+- **[R-IBC-client]** create and update IBC clients on destination chains 
+- **[R-accounts]** Own accounts on destination chains with sufficient balance to pay for transaction fees
+- **[R-transact]** Create, sign and forward IBC datagram transactions
+- **[R-relay]** Perform correct relaying of all required messages, according to the IBC sub-protocol constraints
+- **[R-restart]** Resume correct functionality after restarts
+- **[R-upgrade]** Resume correct functionality after upgrades
+- **[R-proofs]** Perform proof verification (as it will be done on the destination chain) and not forward messages where proof verification fails
+
+The relayer MAY:
+- **[R-config-cli]** Provide ways to change configuration at runtime
+- **[R-bisection]** Perform bisection to optimize transaction costs and computation on destination chains
+- **[R-relay-prio]** Filter or order transactions based on some criteria (e.g. in accordance with the fee payment model)
+
+## Implementation
+The initial implementation will heavily borrow from the Go relayer implementation that uses a "naive" algorithm for relaying messages. The structure of the configuration file is similar with the one in Go (see [Go-Relayer](https://github.com/cosmos/relayer))
+
+### Configuration
+(WIP)
+Upon start the relayer reads a configuration file that includes global and per chain parameters. The file format is .toml
+The relayer performs initialization based on the content of this file. Below is an example of a configuration file.
+
+```$xslt
+# This is an IBC relayer sample configuration
+title = "IBC Relayer Config Example"
+
+[global]
+timeout = "10s"
+strategy = "naive"
+
+[[chains]]
+  id = "chain_A"
+  rpc_addr = "http://localhost:26657"
+  account_prefix = "cosmos"
+  key_name = "testkey"
+  store_prefix = "ibc"
+  client_ids = ["clA1", "clA2"]
+  gas = 200000
+  gas_adjustement = 1.3
+  gas_price = "0.025stake"
+  trusting_period = "336h"
```

Comment:
It might be more clear (and clean) if parameters are grouped (for example light client related goes under light client section, chain related under chain section, etc).

Reply:
Yes, the config file will change quite a bit probably. We will have an issue/PR just for this.